cfd_protocol/src/protocol.rs

```diff
@@ -0,0 +1,532 @@
+pub use transaction_ext::TransactionExt;
+pub use transactions::punish_transaction;
+
+use crate::interval;
+use crate::protocol::sighash_ext::SigHashExt;
+use crate::protocol::transactions::{
+    lock_transaction, CommitTransaction, ContractExecutionTransaction as ContractExecutionTx,
+    RefundTransaction,
+};
+
+use anyhow::{bail, Context, Result};
+use bdk::bitcoin::hashes::hex::ToHex;
+use bdk::bitcoin::util::bip143::SigHashCache;
+use bdk::bitcoin::util::psbt::PartiallySignedTransaction;
+use bdk::bitcoin::{Address, Amount, PublicKey, SigHashType, Transaction, TxOut};
+use bdk::database::BatchDatabase;
+use bdk::descriptor::Descriptor;
+use bdk::miniscript::descriptor::Wsh;
+use bdk::miniscript::DescriptorTrait;
+use bdk::wallet::AddressIndex;
+use bdk::FeeRate;
+use itertools::Itertools;
+use secp256k1_zkp::{self, schnorrsig, EcdsaAdaptorSignature, SecretKey, Signature, SECP256K1};
+use std::collections::HashMap;
+use std::iter::FromIterator;
+
+mod sighash_ext;
+mod transaction_ext;
+mod transactions;
+
+/// Static script to be used to create lock tx
+const DUMMY_2OF2_MULTISIG: &str =
+    "0020b5aa99ed7e0fa92483eb045ab8b7a59146d4d9f6653f21ba729b4331895a5b46";
+
+pub trait WalletExt {
+    fn build_party_params(&self, amount: Amount, identity_pk: PublicKey) -> Result<PartyParams>;
+}
+
+impl<B, D> WalletExt for bdk::Wallet<B, D>
+where
+    D: BatchDatabase,
+{
+    fn build_party_params(&self, amount: Amount, identity_pk: PublicKey) -> Result<PartyParams> {
+        let mut builder = self.build_tx();
+        builder
+            .ordering(bdk::wallet::tx_builder::TxOrdering::Bip69Lexicographic)
+            .fee_rate(FeeRate::from_sat_per_vb(1.0))
+            .add_recipient(
+                DUMMY_2OF2_MULTISIG.parse().expect("Should be valid script"),
+                amount.as_sat(),
+            );
+        let (lock_psbt, _) = builder.finish()?;
+        let address = self.get_address(AddressIndex::New)?.address;
+        Ok(PartyParams {
+            lock_psbt,
+            identity_pk,
+            lock_amount: amount,
+            address,
+        })
+    }
+}
+
+pub fn create_cfd_transactions(
+    (maker, maker_punish_params): (PartyParams, PunishParams),
+    (taker, taker_punish_params): (PartyParams, PunishParams),
+    (oracle_pk, nonce_pks): (schnorrsig::PublicKey, &[schnorrsig::PublicKey]),
```

Comment:
The slice here represents the nonces per digit yes? I think we will at least need to documentation on whether that is in big or little endian order or create a data structure that encodes this somehow.